deps/fast_float_c_interface/valkey_strtod.h

```diff
@@ -0,0 +1,82 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *   * Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of Redis nor the names of its contributors may be used
+ *     to endorse or promote products derived from this software without
+ *     specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#ifndef FAST_FLOAT_STRTOD_H
+#define FAST_FLOAT_STRTOD_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifdef USE_FAST_FLOAT
+
+/**
+ * Converts a null-terminated byte string to a double using the fast_float library.
+ *
+ * This function provides a C-compatible wrapper around the fast_float library's string-to-double
+ * conversion functionality. It aims to offer a faster alternative to the standard strtod function.
+ *
+ * str: A pointer to the null-terminated byte string to be converted.
+ * value: A pointer to the double variable where the function stores converted double value.
+ *        On success, the function stores the converted double value. On failure, it stores 0.0 and
+ *        stores error code in errno to ERANGE or EINVAL.
+ *
+ * return value: On success, returns char pointer pointing to '\0' at the end of the string.
+ *               On failure, returns char pointer pointing to first invalid character in the string.
+ *
+ * note: This function uses the fast_float library (https://github.com/fastfloat/fast_float) for
+ * the actual conversion, which can be significantly faster than standard library functions.
+ *
+ * Refer to https://github.com/fastfloat/fast_float for more information on the underlying library.
+ */
+const char *fast_float_strtod(const char *str, double *value);
+
+static inline const char *valkey_strtod(const char *str, double *value) {
+    errno = 0;
+    return fast_float_strtod(str, value);
+}
```

Comment:
1. I'm fine with the directory name `fast_float_c_interface`, although it's not what we agreed on in the long chat. We agreed on `fast_float_inteface`. If we keep changing things after making decisions, we have to make new decisions over and over.

2. This file provides a static inline `valkey_strtod` function and it's located under `deps/fast_float_c_interface/`, but it doesn't always use fast float. It only conditionally uses it.

   The purpose of naming the wrapper `valkey_strtod` rather than `fast_float_strtod` was that it should be implementation agnostic to the caller. When it's placed in this directory, it isn't. I see that files needing strtod do `#include "../deps/fast_float_c_interface/valkey_strtod.h` to get the `valkey_strtod` function.

   We have three pieces here:

   1. the external fast float C++ header (`deps/fast_float/fast_float.h`)
   2. interface code to be able to call C++ from C. IMO this should be specific to fast float and be called something like `deps/fast_float_c_interface/fast_float_strtod.{cpp,h}`, but @madolson you don't like this name?
   3. conditional code to choose between the fast float and the stdlib implementations. IMO this should be placed in `src/valkey_strtod.h`.

   For iii, compare this to jemalloc. The conditional code for which allocator we use is abstracted by zmalloc and is placed in `src/zmalloc.{c,h}`.

   If you think differently, please convince me. I may be wrong again. :)

Reply:
@zuiderkwast I was expecting this file to end up under src/*, since as you mentioned its our code. This is not CPP code.

Reply:
It seems we agree to use `deps/fast_float_c_interface/fast_float_strtod.cpp` for the interfacing code and `src/valkey_strtod.h` for our conditional code then?

Reply:
Yes, I agree with that.

Reply:
Moving valkey_strtod.h to src/.